codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoMiddlewareGenerator.java

```diff
@@ -0,0 +1,201 @@
+        this.handlerType = builder.handlerType;
+    }
+
+    public static GoMiddlewareGenerator newSerializeMiddleware(String identifier) {
+        return builder()
+                .identifier(identifier)
+                .handleMethodName("HandleSerialize")
+                .inputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.SerializeInput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .outputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.SerializeOutput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .handlerType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.SerializeHandler")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .build();
+    }
+
+    public static GoMiddlewareGenerator newDeserializeMiddleware(String identifier) {
+        return builder()
+                .identifier(identifier)
+                .handleMethodName("HandleDeserialize")
+                .inputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeInput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .outputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeOutput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .handlerType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeHandler")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .build();
+    }
+
+    public void writeMiddleware(GoWriter writer, BiConsumer<GoMiddlewareGenerator, GoWriter> handlerBodyConsumer) {
+        writeMiddleware(writer, (m, w) -> {
+        }, handlerBodyConsumer);
+    }
+
+    public void writeMiddleware(GoWriter writer, BiConsumer<GoMiddlewareGenerator, GoWriter> fieldConsumer,
```

Comment:
Sadly we don't have something as nice as go fmt, but when args make a line longer than 120 characters you should just put one arg per line, e.g.:

```java
public void foo(
    String bar,
    String baz
) {
    // stuff
}
```

I would also recommend swapping the order of `fieldConsumer` and `handlerBodyConsumer` as it's a bit more ergonomic for optional parameters to be at the end.